tonic/src/service/layered.rs

```diff
@@ -0,0 +1,48 @@
+use std::{
+    marker::PhantomData,
+    task::{Context, Poll},
+};
+
+use tower_layer::Layer;
+use tower_service::Service;
+
+use crate::server::NamedService;
+
+/// A layered service to propagate [`NamedService`] implementation.
+#[derive(Debug, Clone)]
+pub struct Layered<S, T> {
+    inner: S,
+    _ty: PhantomData<T>,
+}
+
+/// Create a new service from the service and the layer.
+pub fn layered_service<S, L>(service: S, layer: L) -> Layered<L::Service, S>
```

Comment:
It feels more idiomatic to me to make rename this to `Layered::new()`, but I'm not very familiar with the conventions in the tower ecosystem.

Reply:
The current design is a result of the type system. It can also be the constructor of the associated function, but in that case `Layered` requires one more type parameter to constrain the type of the `Layer`.

Reply:
Not sure I understand this. Can you demonstrate?

Reply:
Added a commit to show it.

Reply:
Considering that a common pattern for applying a layer to a service to create a new service is to use `layer.layer(service)` rather than providing a constructor for the service, it might be more appropriate to provide a trait that wraps the service with Layered when applying a layer to the service rather than providing a constructor.